core/src/test/java/org/infinispan/remoting/RemoteShutdownDuringOperationTest.java

```diff
@@ -0,0 +1,195 @@
+            .getComponent(ClusterTopologyManager.class);
+
+      try {
+         CheckPoint checkPoint = new CheckPoint();
+         CheckPointInterceptor interceptor = new CheckPointInterceptor(checkPoint);
+         extractInterceptorChain(c1).addInterceptor(interceptor, 1);
+
+         ClusterTopologyManager spyCtm = spy(ctm);
+         CompletableFuture<Void> cf = new CompletableFuture<>();
+         doAnswer(ivk -> {
+            // We delay the handling of the leave request until *after* the command is handled.
+            cf.thenRunAsync(() -> {
+               try {
+                  ivk.callRealMethod();
+               } catch (Throwable e) {
+                  fail(e.getMessage());
+               }
+            }, testExecutor());
+            return CompletableFutures.completedNull();
+         }).when(spyCtm).handleLeave(eq(CACHE_NAME), any(), anyInt());
+         TestingUtil.replaceComponent(manager(0), ClusterTopologyManager.class, spyCtm, true);
+
+         SoftAssertions softly = new SoftAssertions();
+
+         CompletableFuture<Void> cs = operation.operate(c0, key, softly).toCompletableFuture();
+
+         // Wait for the remote node to receive the command and shutdown it.
+         checkPoint.awaitStrict("receive_remote_command", 10, TimeUnit.SECONDS);
+         c1.stop();
+         assertThat(c1.getAdvancedCache().getComponentRegistry().getStatus().isTerminated()).isTrue();
+
+         // Proceed handling the command, which will fail with a CacheNotFoundResponse.
+         checkPoint.trigger("proceed_remote_command", 1);
+
+         // Now that the command was handled, process the leave request.
+         checkPoint.awaitStrict("response_remote_command", 10, TimeUnit.SECONDS);
+         cf.complete(null);
+
+         // Wait for the command to complete.
+         cs.get(10, TimeUnit.SECONDS);
+         softly.assertAll();
+      } finally {
+         TestingUtil.replaceComponent(manager(0), ClusterTopologyManager.class, ctm, true);
+      }
+   }
+
+   @Override
+   protected EncoderCache<MagicKey, String> cache(int index) {
+      return (EncoderCache<MagicKey, String>) (Object) cache(index, CACHE_NAME);
+   }
+
+   @Override
+   public Object[] factory() {
+      return Stream.of(TolerateSuspectOperation.values())
+            .map(op -> new RemoteShutdownDuringOperationTest().withOperation(op))
+            .toArray();
+   }
+
+   @Override
+   protected String parameters() {
+      return "[operation=" + operation + "]";
+   }
+
+   static class CheckPointInterceptor extends DDAsyncInterceptor {
+      private final CheckPoint checkPoint;
+
+      CheckPointInterceptor(CheckPoint checkPoint) {
+         this.checkPoint = checkPoint;
+      }
+
+      @Override
+      public Object visitGetCacheEntryCommand(InvocationContext ctx, GetCacheEntryCommand command) throws Throwable {
+         checkPoint.trigger("receive_remote_command", 1);
+         checkPoint.awaitStrict("proceed_remote_command", 10, TimeUnit.SECONDS);
```

Comment:
This can use checkPoint.future instead to not block the caller btw

Reply:
Somehow I didn't notice that method before :sweat_smile:. Updated!